allow Controller to be focus while there is an error

bluebill1049 opened the pull request:
This issue has bothered me for awhile: https://github.com/react-hook-form/react-hook-form/issues/1237 it makes Controller autoFocus terrible for DX, to a point when @JuanmaMenendez was suggested to disable that feature all together. here i would like to propose this solution and codesandbox below:

https://codesandbox.io/s/react-hook-form-controller-test-out-ref-5tru5

```
<Controller
  as={<TextField inputRef={inputRef} />}
  name="TextField"
  onFocus={() => inputRef.current.focus()}
  control={control}
  rules={{ required: true }}
/>
```

bluebill1049 committed:
allow Controller to be focus while there is an error

bluebill1049 committed:
fix type issue

JeromeDeLeon reviewed (dismissed):
LGTM! This will only be triggered upon validation error therefore, it needs to be documented as some people might assume that `onFocus` triggers upon focus as the name implies.

JeromeDeLeon commented on src/logic/focusErrorField.ts:
```diff
@@ -10,8 +9,8 @@ export default <FormValues>(
-        if (isHTMLElement(field.ref) && field.ref.focus) {
-          field.ref.focus();
+        if ((field.ref as FieldElement).focus) {
+          (field.ref as any).focus();
```
Is `any` really necessary? I just tried it with the csb and nothing wrongs if we omit `any` here.

bluebill1049 replied:
TS cries about it, can you try out in your env?

bluebill1049 replied:
feel free to commit on top of my branch of you have better suggestion 🛶

JeromeDeLeon replied:
will do.

JeromeDeLeon committed:
Remove type

Yes, there was no error on my end. I also tried it earlier on the csb your provided.

JeromeDeLeon committed:
Remove unused declaration

bluebill1049 commented:
awesome! thanks @JeromeDeLeon!

JuanmaMenendez commented:
@bluebill1049  the "autoFocus" on error is not working for me in your https://codesandbox.io/s/react-hook-form-controller-test-out-ref-5tru5. Where can I test it?

JeromeDeLeon commented:
Try applying the changes to that csb manually to see if it works.

bluebill1049 commented:
https://codesandbox.io/s/react-hook-form-controller-test-out-ref-5tru5 try this

JuanmaMenendez commented:
Thanks @bluebill1049 it works now. 
It Will be great to make it work also with the Autocomplete Component from MUI which has to use Controller and it is very used. I tried to make it work with your technique but without success, maybe you can make it work?
https://codesandbox.io/s/mui-autocomplete-with-react-hook-form-ljttt

bluebill1049 commented:
@JuanmaMenendez as long as you can hold off the ref of that input ref, you can focus them. However, if MUI is not forward the ref, then there is very little we can do.

JuanmaMenendez commented:
I understand but it should work, I am trying to focus a TextField, the same element you used in your example.

Also, the Autocomplete component which makes use of a TextField also for forward the ref, you can read it here close to the end  https://material-ui.com/api/autocomplete/#autocomplete-api    "The ref is forwarded to the root element."

bluebill1049 commented:
@JuanmaMenendez this feature is not merged, you need to fork my CSB above.

JuanmaMenendez commented:
🤦‍♂️ 

Now I tried forking you CSB and I replaced your index.js and add App.js and Countries.js but still without luck  https://codesandbox.io/s/react-hook-form-controller-test-out-ref-jtswf

bluebill1049 commented:
<img width="777" alt="Screen Shot 2020-03-30 at 10 40 47 am" src="https://user-images.githubusercontent.com/10513364/77864167-f0cc9900-7272-11ea-943f-4ccc6dc36480.png">

you are still pointing to a live version. 😂

bluebill1049 commented:
let me have a play and get you a version.

bluebill1049 commented:
There you go: https://codesandbox.io/s/react-hook-form-controller-test-out-ref-7imfk make you point to ./src folder not 'react-hook-form'.

JuanmaMenendez commented:
haha my bad, thanks @bluebill1049 it works perfectly, very good job! I will update the issue where we have the discussion. 

Maybe you can add me as a Contributor and I can help you guys sometimes.

rmlevangelio commented:
Lol love this. I needed this before but I decided to convert my custom component to use ref. XD

bluebill1049 commented:
> haha my bad, thanks @bluebill1049 it works perfectly, very good job! I will update the issue where we have the discussion.
> 
> Maybe you can add me as a Contributor and I can help you guys sometimes.

Thanks @JuanmaMenendez you are already a contributor for this lib by raising the issue, testing and verifying the solution. 👍

bluebill1049 commented:
once we get a green light from @kotarella1110 let's merge this 🥂

bluebill1049 commented:
> Lol love this. I needed this before but I decided to convert my custom component to use ref. XD

better late than never 😅

kotarella1110 reviewed (approved):
LGTM! This is a feature that many people using RHF need 💚

sanohin commented:
What if we pass ref to a component and allow a user to enable that by passing `inputRefName` prop? This will allow us to avoid adding refs in our forms code

bluebill1049 commented:
@sanohin thanks for your suggestion. I think libs have different ref implementation as far as I know, I think it's safer to do assign the ref by the users.

sanohin commented:
@bluebill1049 ok, thank you, when do you plan to release the new version of react-hook-form with this improvement?

bluebill1049 commented:
We should be able to get this one out the door by this weekend. @sanohin

Falkovic1993 commented:
Hey @bluebill1049 thanks for this! This is awesome - do you have a solution for etc. radiobuttons? 

Eg. 

<Controller 
                        as={
                        <RadioGroup row>
                          <FormControlLabel  value="delivery" control={<Radio />} label="test"/>
                          <FormControlLabel  value="pickup" control={<Radio />} label="Pickup yourself" />
                          <FormControlLabel value="eatin" control={<Radio />} label="Eat at restaurant" />
                        </RadioGroup>
                        }
                        onChange={([e]) => {
                          e.persist();
                          handleDeliveryMethodChange(e)
                        }}
                        rules={{ required: true }}
                        name="delivery"
                        control={control}
                        />

It works perfectly for text fields, but having some problem making it work on this. Thanks once again!

bluebill1049 commented:
I would get all three radio input ref and focus on the one have error inside `onfocus` function.
